docs/user/exercises/programming-exercise-AI-proof-guidelines.inc

```diff
@@ -0,0 +1,491 @@
+      :alt: This figure showcases the functionality of a linked-list
+      :width: 50%
+
+
+      Depiction shows the functionality of a linked list with 3 integer values
+      as a basic concept for understanding linked lists. The task of this exercise
+      were to implement the common List functionalities from scratch rather than
+      rely on the standard methods from java. ListNode and LinkedList were
+      therefore own classes that needed to be filled with the required
+      functionalities.
+
+
+   This depiction provides an understanding of how a linked list should look
+   like after the add function was called.
+
+   .. figure:: programming/AI-proof-exercises/add.png
+      :align: center
+      :width: 50%
+
+
+      Figure shows what a list should look like after the add functionality was called. Picture from the Problem statement
+
+-  **Text in picture format:** Giving the text exercise instructions in
+   a picture format that can not be copied. Rather than writing long
+   inquiries into the AI, the students will be promoted to asking
+   reasonable questions and solve exercises on their own.
+
+   .. note::
+      Since there are already possibilities to copy text from pictures, the effect of
+      this measure are limited. More suitable for exam mode.
+
+-  **Java frameworks:** Using Java frameworks, such as JavaFX, is challenging for AI
+   and helps to visualize projects in a more interactive way
+   for students through gamification. How can we incorporate that for
+   our exercises?
+
+   -  Provide Graphical User Interfaces (GUI) for
+      students with which they can also test the functionality of their
+      code.
+
+   -  Give exercises that use GUIs because AI has difficulty
+      solving them accurately.
+
+   -  Provide pictures of how the GUI should look in the end.
+
+      Students needed to implement the GUI based on the this figure that
+      was given in the problem statement.
+
+      .. figure:: programming/AI-proof-exercises/GUI-upgrade-for-game.png
+         :align: center
+         :width: 30%
+
+
+         Example GUI from homework exercise
+
+   .. note::
+      Apply this approach when students already have knowledge of what a GUI is. Most likely in
+      a later stage of the semester.
+
+.. note::
+   ChatGPT is limited to text input. Including a fair amount of graphical
+   representation for the exercises helps to limit the effectiveness of such tools.
+
+Problem statement and exercise instructions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+
+-  **Exercise length:** There is a negative correlation between the
+   length of exercises and AI performance:
+
+   -  Do not keep the exercises between 500 and 1000 characters, but
+      rather keep them between 1200 and 2500 characters. Short exercises
+      are normally easier to solve for AI since there are not so many
+      constraints and the problem statement only focused on the most
+      relevant implementation parts.
```

Comment:
This is somewhat true, but only partially. During the development of Iris we could not really see a big difference between problem statement length and its ability to (despite our intial safeguards) generate correct code.

Also making the problem statement longer is a really hard thing to do without just making it overly verbose and hard to read.

And ultimately you could also just preface any prompting by first using an LLM to shorten the problem statement to the essential parts.

Reply:
I got that from a study where they tested the correlation between input length and ability to solve exercises. This recommendations should maybe just be a general guidline on how long exercises should be. Obviously that is not possible for the first few exercises where basic print statements are asked from students. Creators need to see where it makes sense, i guess.

Reply:
Also this negative correlation is only temporary. GPT4 has a longer context length it can take before deteriorating than GPT3.5. So this is also something that will age fast, as you cannot make the problem statement arbitrarily long.